lib/utmp.c

```diff
@@ -156,24 +156,27 @@ failtmp(const char *username, const struct utmpx *failent)
 static /*@null@*/ /*@only@*/struct utmpx *
 get_current_utmp(pid_t main_pid)
 {
-	struct utmpx  *ut;
-	struct utmpx  *ret = NULL;
-	struct utmpx  *ut_by_pid  = NULL; /* first entry matched by 'ut_pid' */
-	struct utmpx  *ut_by_line = NULL; /* first entry matched by 'ut_line' */
+	struct utmpx *ut;
+	struct utmpx *ut_by_pid  = NULL; /* first entry matched by 'ut_pid' */
+	struct utmpx *ut_by_line = NULL; /* first entry matched by 'ut_line' */
 
 	setutxent();
 
-	while ((ut = getutxent()) != NULL) {
-		if (   (LOGIN_PROCESS == ut->ut_type)
-		    || (USER_PROCESS  == ut->ut_type)) {
-			if (main_pid == ut->ut_pid) {
-				if (is_my_tty(ut->ut_line))
-					break; /* Found the entry */
-				if (NULL == ut_by_pid)
-					ut_by_pid = ut;  /* Matched only by 'ut_pid' */
-			} else if ((NULL == ut_by_line) && is_my_tty(ut->ut_line))
-				ut_by_line = ut; /* Matched only by 'ut_line' */
-		}
+	for (ut = getutxent(); ut != NULL; ut = getutxent()) {
+		if (   (LOGIN_PROCESS != ut->ut_type)
+		    && (USER_PROCESS  != ut->ut_type))
+			continue; /* Look for 'login' or 'user' entries only */
+
+		if (main_pid == ut->ut_pid) {
+			if (is_my_tty(ut->ut_line))
+				break; /* Found the entry */
+
+			if (NULL == ut_by_pid)
+				ut_by_pid = ut;  /* Matched only by 'ut_pid' */
+
+		} else if (   (NULL == ut_by_line)
+				   && is_my_tty(ut->ut_line))
+			ut_by_line = ut; /* Matched only by 'ut_line' */
 	}
 
```

Comment:
I would prefer the refactor to go before the actual change.  That will result in both a smaller refactor and a smaller change.